File: 2025-9322-team04_finproject_java/src/main/java/Client_Java/player/PlayerCorbaManager.java
package Client_Java.player;

import Server_Java.manager.DatabaseManager;
import compilations.*;
import org.omg.CORBA.ORB;
import org.omg.CosNaming.NamingContextExt;
import org.omg.CosNaming.NamingContextExtHelper;

import java.io.InputStream;
import java.sql.Connection;
import java.util.Properties;

public class PlayerCorbaManager {
    private static PlayerCorbaManager instance;
    private static Connection connection;

    private ORB orb;
    private NamingContextExt namingContext;

    // Service references
    private LoginService loginService;
    private GameService gameService;
    private AdminService adminService;
    private LeaderboardService leaderboardService;
    private WordService wordService;
    private GameManagerService gameManagerService;

    private PlayerCorbaManager(String[] orbArgs) {
        try {
            Properties orbProps = new Properties();
            try (InputStream input = PlayerCorbaManager.class.getClassLoader().getResourceAsStream("config.properties")) {
                if (input != null) {
                    orbProps.load(input);
                } else {
                    System.err.println("PlayerCorbaManager: config.properties not found. Using default ORB settings.");
                }
            } catch (Exception e) {
                System.err.println("PlayerCorbaManager: Failed to load config.properties: " + e.getMessage());
            }

            orb = ORB.init(orbArgs, orbProps);

            // Resolve naming service
            org.omg.CORBA.Object objRef = orb.resolve_initial_references("NameService");
            namingContext = NamingContextExtHelper.narrow(objRef);

            // Connect to services
            loginService = LoginServiceHelper.narrow(namingContext.resolve_str("LoginService"));
            gameService = GameServiceHelper.narrow(namingContext.resolve_str("GameService"));
            adminService = AdminServiceHelper.narrow(namingContext.resolve_str("AdminService"));
            leaderboardService = LeaderboardServiceHelper.narrow(namingContext.resolve_str("LeaderboardService"));
            wordService = WordServiceHelper.narrow(namingContext.resolve_str("WordService"));
            gameManagerService = GameManagerServiceHelper.narrow(namingContext.resolve_str("GameManagerService"));

            // Database connection
            connection = DatabaseManager.getConnection();

            System.out.println("PlayerCORBAManager: Connected to all services.");
        } catch (Exception e) {
            System.err.println("PlayerCORBAManager: Failed to connect to services.");
            e.printStackTrace();
        }
    }

    // First-time initializer
    public static PlayerCorbaManager getInstance(String[] orbArgs) {
        if (instance == null) {
            synchronized (PlayerCorbaManager.class) {
                if (instance == null) {
                    instance = new PlayerCorbaManager(orbArgs);
                }
            }
        }
        return instance;
    }

    // Getter (after initialized)
    public static PlayerCorbaManager getInstance() {
        if (instance == null) {
            throw new IllegalStateException("CORBAConnector not initialized. Call getInstance(String[] args) first.");
        }
        return instance;
    }

    public Connection getDbConnection() {
        return connection;
    }

    private String currentPlayerUsername;
    private String sessionId;

    public String getCurrentPlayerUsername() {
        return currentPlayerUsername;
    }

    public void setCurrentPlayerUsername(String username) {
        this.currentPlayerUsername = username;
    }

    public String getSessionId() {
        return sessionId;
    }

    public void setSessionId(String sessionId) {
        this.sessionId = sessionId;
    }

    // Service accessors
    public LoginService getLoginService() {
        return loginService;
    }

    public GameService getGameService() {
        return gameService;
    }

    public AdminService getAdminService() {
        return adminService;
    }

    public LeaderboardService getLeaderboardService() {
        return leaderboardService;
    }

    public WordService getWordService() {
        return wordService;
    }

    public GameManagerService getGameManagerService() {
        return gameManagerService;
    }

    /**
     * Re-establishes CORBA connection and reinitializes all services
     */
    public synchronized void reconnect(String[] orbArgs) {
        try {
            System.out.println("PlayerCORBAManager: Attempting to reconnect CORBA services...");

            // Reinitialize ORB with provided parameters
            orb = ORB.init(orbArgs, null);

            // Re-resolve naming service
            org.omg.CORBA.Object objRef = orb.resolve_initial_references("NameService");
            namingContext = NamingContextExtHelper.narrow(objRef);

            // Reconnect to all services
            loginService = LoginServiceHelper.narrow(namingContext.resolve_str("LoginService"));
            gameService = GameServiceHelper.narrow(namingContext.resolve_str("GameService"));
            adminService = AdminServiceHelper.narrow(namingContext.resolve_str("AdminService"));
            leaderboardService = LeaderboardServiceHelper.narrow(namingContext.resolve_str("LeaderboardService"));
            wordService = WordServiceHelper.narrow(namingContext.resolve_str("WordService"));
            gameManagerService = GameManagerServiceHelper.narrow(namingContext.resolve_str("GameManagerService"));

            System.out.println("PlayerCORBAManager: CORBA services reconnected successfully");
        } catch (Exception e) {
            System.err.println("PlayerCORBAManager: CORBA reconnection failed: " + e);
            e.printStackTrace();
        }
    }

    /**
     * Force a fresh initialization of all services
     * @param orbArgs ORB initialization arguments
     */
    public synchronized void reinitialize(String[] orbArgs) {
        try {
            System.out.println("PlayerCORBAManager: Performing full CORBA reinitialization...");

            // Shutdown existing ORB if present
            if (orb != null) {
                try {
                    orb.shutdown(false);
                } catch (Exception e) {
                    System.err.println("PlayerCORBAManager: Error shutting down existing ORB: " + e);
                }
            }

            // Create completely new instance
            instance = new PlayerCorbaManager(orbArgs);
            System.out.println("PlayerCORBAManager: CORBA services fully reinitialized");
        } catch (Exception e) {
            System.err.println("PlayerCORBAManager: Full reinitialization failed: " + e);
            e.printStackTrace();
        }
    }
}

File: 2025-9322-team04_finproject_python/Client_Python/manager/PlayerCorbaManager.py
import sys
import os
import threading
import traceback

import CORBA
import CosNaming

# Ensure the parent directory is in the path for imports
sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), '..')))

# CORBA IDL-generated stubs
from Client_Python.compilations import LoginService_idl
from Client_Python.compilations import GameService_idl
from Client_Python.compilations import AdminService_idl
from Client_Python.compilations import LeaderboardService_idl
from Client_Python.compilations import WordService_idl
from Client_Python.compilations import GameManagerService_idl


class PlayerCorbaManager:
    _instance = None
    _lock = threading.Lock()

    def __init__(self, orb_args):
        try:
            # Initialize ORB
            self.orb = CORBA.ORB_init(orb_args, "")

            # Resolve naming service
            obj_ref = self.orb.resolve_initial_references("NameService")
            self.naming_context = obj_ref._narrow(CosNaming.NamingContextExt)

            # Resolve and connect to all services using _0_compilations._objref_*
            obj = self.naming_context.resolve_str("LoginService")
            self.login_service = obj._narrow(LoginService_idl._0_compilations.LoginService)
            self._validate_service(self.login_service, "LoginService")
            print()

            obj = self.naming_context.resolve_str("GameService")
            self.game_service = obj._narrow(GameService_idl._0_compilations.GameService)
            self._validate_service(self.game_service, "GameService")

            obj = self.naming_context.resolve_str("AdminService")
            self.admin_service = obj._narrow(AdminService_idl._0_compilations.AdminService)
            self._validate_service(self.admin_service, "AdminService")

            obj = self.naming_context.resolve_str("LeaderboardService")
            self.leaderboard_service = obj._narrow(LeaderboardService_idl._0_compilations.LeaderboardService)
            self._validate_service(self.leaderboard_service, "LeaderboardService")

            obj = self.naming_context.resolve_str("WordService")
            self.word_service = obj._narrow(WordService_idl._0_compilations.WordService)
            self._validate_service(self.word_service, "WordService")

            obj = self.naming_context.resolve_str("GameManagerService")
            self.game_manager_service = obj._narrow(GameManagerService_idl._0_compilations.GameManagerService)
            self._validate_service(self.game_manager_service, "GameManagerService")

            self.current_player_username = None
            print("✔ PlayerCorbaManager: Connected to all services.")

        except Exception as e:
            print("X PlayerCorbaManager: Failed to connect to services.", file=sys.stderr)
            traceback.print_exc()

    @classmethod
    def get_instance(cls, orb_args=None):
        with cls._lock:
            if cls._instance is None:
                if orb_args is None:
                    raise ValueError("orb_args required for first initialization")
                cls._instance = cls(orb_args)
        return cls._instance

    def _validate_service(self, service, name):
        if service is None:
            raise RuntimeError(f"{name} is not available via CORBA naming service.")

    # Username management
    def set_current_player_username(self, username):
        #print("set_current_player_username", username)
        self.current_player_username = username

    def get_current_player_username(self):
        #print("get_current_player_username")
        return self.current_player_username

    # Service getters
    def get_login_service(self):
        #print("get_login_service")
        return self.login_service

    def get_game_service(self):
        #print("get_game_service")
        return self.game_service

    def get_admin_service(self):
        #print("get_admin_service")
        return self.admin_service

    def get_leaderboard_service(self):
        #print("get_leaderboard_service")
        return self.leaderboard_service

    def get_word_service(self):
        #print("get_word_service")
        return self.word_service

    def get_game_manager_service(self):
        #print("get_game_manager_service")
        return self.game_manager_service

    def get_session_id(self):
        return self.session_id

    def set_session_id(self, session_id):
        self.session_id = session_id

    # Optional ORB shutdown method
    def shutdown(self):
        try:
            self.orb.shutdown(wait=False)
            print("ORB shutdown completed.")
        except Exception as e:
            print("Failed to shut down ORB:", e)
